discovery/scaleway/scaleway.go

```diff
@@ -0,0 +1,200 @@
+	// If you need to configure multiple zones use multiple scaleway_sd_configs
+	Zone string `yaml:"zone"`
+	// AccessKey used to authenticate on Scaleway APIs.
+	AccessKey string `yaml:"access_key"`
+	// SecretKey used to authenticate on Scaleway APIs.
+	SecretKey config.Secret `yaml:"secret_key"`
+	// NameFilter to filter on during the ListServers.
+	NameFilter string `yaml:"name_filter,omitempty"`
+	// TagsFilter to filter on during the ListServers.
+	TagsFilter []string `yaml:"tags_filter,omitempty"`
+
+	HTTPClientConfig config.HTTPClientConfig `yaml:",inline"`
+
+	RefreshInterval model.Duration `yaml:"refresh_interval"`
+	Port            int            `yaml:"port"`
+	// Role can be either instance or baremetal
+	Role role `yaml:"role"`
+}
+
+func (c SDConfig) Name() string {
+	return "scaleway"
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+
+	if c.Role == "" {
+		return errors.New("role missing (one of: instance, baremetal)")
+	}
+
+	if c.Project == "" {
+		return errors.New("project_id is mandatory")
+	}
+
+	if c.SecretKey == "" {
+		return errors.New("secret_key is mandatory")
+	}
+
+	if c.AccessKey == "" {
+		return errors.New("access_key is mandatory")
+	}
+
+	profile, err := loadProfile(c)
+	if err != nil {
+		return err
+	}
+	_, err = scw.NewClient(
+		scw.WithProfile(profile),
+	)
+	if err != nil {
+		return err
+	}
+
+	return c.HTTPClientConfig.Validate()
+}
+
+func (c SDConfig) NewDiscoverer(options discovery.DiscovererOptions) (discovery.Discoverer, error) {
+	return NewDiscovery(&c, options.Logger)
+}
+
+// SetDirectory joins any relative file paths with dir.
+func (c *SDConfig) SetDirectory(dir string) {
+	c.HTTPClientConfig.SetDirectory(dir)
+}
+
+func init() {
+	discovery.RegisterConfig(&SDConfig{})
+}
+
+// Discovery periodically performs Scaleway requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+}
+
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*refresh.Discovery, error) {
+	r, err := newRefresher(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	return refresh.NewDiscovery(
+		logger,
+		"scaleway",
+		time.Duration(conf.RefreshInterval),
+		r.refresh,
+	), nil
+}
+
+type refresher interface {
+	refresh(context.Context) ([]*targetgroup.Group, error)
+}
+
+func newRefresher(conf *SDConfig) (refresher, error) {
+	switch conf.Role {
+	case roleBaremetal:
+		return newBaremetalDiscovery(conf)
+	case roleInstance:
+		return newInstanceDiscovery(conf)
+	}
+	return nil, errors.New("unknown Scaleway discovery role")
+}
+
+func loadProfile(sdConfig *SDConfig) (*scw.Profile, error) {
+	// Profile coming from Prometheus Configuration file
+	prometheusConfigProfile := &scw.Profile{
+		DefaultZone:      scw.StringPtr(sdConfig.Zone),
+		APIURL:           scw.StringPtr(sdConfig.APIURL),
```

Comment:
nit: when `sdconfig.APIURL` is not set we could use regionalized URLs here based on DefaultZone.
```go
defaultRegion, err := defaultZone.Region()
// check err
apiURL := "api-"+defaultRegion.String()+".scaleway.com"
```

e.g.: `api-fr-par.scaleway.com`, `api-pl-waw.scaleway.com`,...

Reply:
I'd rather keep this simple, as a user who would put a wrong zone might end up with very strange error messages. I also note that the offical documentation points to api.scaleway.com.